model/src/main/java/org/openremote/model/teltonika/State.java

```diff
@@ -0,0 +1,298 @@
+            attributes.addOrReplace(generatedAttribute);
+        }
+        //Special parameter definitions without being defined in the AVL Parameter List, thanks Teltonika
+
+        //latlng are the latitude-longitude coordinates, also check if it's 0,0, if it is, don't update.
+        if (reported.containsKey("latlng") && !Objects.equals(reported.get("latlng"), "0.000000,0.000000")){
+            try{
+                String latlngString = reported.get("latlng").toString();
+                GeoJSONPoint point = ParseLatLngToGeoJSONObject(latlngString);
+                Attribute<?> attr = new Attribute<>("location", ValueType.GEO_JSON_POINT, point);
+
+                attributes.add(attr);
+            }catch (Exception e){
+                logger.severe("Failed coordinates");
+                logger.severe(e.toString());
+                throw e;
+            }
+        }
+        //Timestamp grabbed from the device.
+        //TODO: It doesn't look like the timestamp is being parsed correctly...
+        if (reported.containsKey("ts")){
+            try{
+                long unixTimestampMillis = Long.parseLong(reported.get("ts").toString());
+                Timestamp deviceTimestamp = Timestamp.from(Instant.ofEpochMilli(unixTimestampMillis));
+                attributes.add(new Attribute<>("lastContact", ValueType.DATE_AND_TIME, deviceTimestamp).setTimestamp(deviceTimestamp.getTime()));
+
+
+                //Update all affected attribute timestamps
+                attributes.forEach(attribute -> attribute.setTimestamp(deviceTimestamp.getTime()));
+            }catch (Exception e){
+                logger.severe("Failed timestamps");
+                logger.severe(e.toString());
+                throw e;
+            }
+
+        }
+
+        // Store data points, allow use for rules, and don't allow user parameter modification, for every attribute parsed
+        try{
+            attributes.forEach(attribute -> attribute.addOrReplaceMeta(
+                            new MetaItem<>(STORE_DATA_POINTS, true),
+                            new MetaItem<>(RULE_STATE, true),
+                            new MetaItem<>(READ_ONLY, true)
+                    )
+            );
+        }catch (Exception e){
+            logger.severe("Failed metaItems");
+            logger.severe(e.toString());
+            throw e;
+        }
+
+        return attributes;
+    }
+
+
+
+    private ValueDescriptor<?> GetAttributeType(TeltonikaParameter parameter) {
+        try{
+            Double.valueOf(parameter.min);
+            Double.valueOf(parameter.max);
+            return ValueType.NUMBER;
+        }catch (NumberFormatException e){
+            return switch (parameter.type) {
+                case "Unsigned", "Signed", "unsigned", "UNSIGNED LONG INT" -> ValueType.NUMBER;
+                default -> ValueType.TEXT;
+            };
+        }
+    }
+
+    private GeoJSONPoint ParseLatLngToGeoJSONObject(String latlngString) {
```

Comment:
Make a note of the format of the data being parses

Reply:
Added javadoc